server/src/main/java/org/elasticsearch/index/mapper/JsonFieldMapper.java

```diff
@@ -0,0 +1,321 @@
+            this.splitQueriesOnWhitespace = splitQueriesOnWhitespace;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            if (omitNorms()) {
+                return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
+            } else {
+                return new NormsFieldExistsQuery(name());
+            }
+        }
+
+        @Override
+        public Query fuzzyQuery(Object value, Fuzziness fuzziness, int prefixLength, int maxExpansions,
+                                boolean transpositions) {
+            throw new UnsupportedOperationException("[fuzzy] queries are not currently supported on [" +
+                CONTENT_TYPE + "] fields.");
+        }
+
+        @Override
+        public Query regexpQuery(String value, int flags, int maxDeterminizedStates,
+                                 MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            throw new UnsupportedOperationException("[regexp] queries are not currently supported on [" +
+                CONTENT_TYPE + "] fields.");
+        }
+
+        @Override
+        public Object valueForDisplay(Object value) {
+            if (value == null) {
+                return null;
+            }
+            BytesRef binaryValue = (BytesRef) value;
+            return binaryValue.utf8ToString();
+        }
+    }
+
+    private final JsonFieldParser fieldParser;
+    private int ignoreAbove;
+
+    private JsonFieldMapper(String simpleName,
+                            MappedFieldType fieldType,
+                            MappedFieldType defaultFieldType,
+                            int ignoreAbove,
+                            Settings indexSettings) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, MultiFields.empty(), CopyTo.empty());
+        assert fieldType.indexOptions().compareTo(IndexOptions.DOCS_AND_FREQS) <= 0;
+
+        this.ignoreAbove = ignoreAbove;
+        this.fieldParser = new JsonFieldParser(fieldType, ignoreAbove);
+    }
+
+    @Override
+    protected String contentType() {
+        return CONTENT_TYPE;
+    }
+
+    @Override
+    protected void doMerge(Mapper mergeWith) {
+        super.doMerge(mergeWith);
+        this.ignoreAbove = ((JsonFieldMapper) mergeWith).ignoreAbove;
+    }
+
+    @Override
+    protected JsonFieldMapper clone() {
+        return (JsonFieldMapper) super.clone();
+    }
+
+    @Override
+    public JsonFieldType fieldType() {
+        return (JsonFieldType) super.fieldType();
+    }
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        if (context.parser().currentToken() == XContentParser.Token.VALUE_NULL) {
+            return;
+        }
+
+        if (fieldType().indexOptions() != IndexOptions.NONE || fieldType().stored()) {
+            fields.addAll(fieldParser.parse(context.parser()));
+            if (fieldType.omitNorms()) {
```

Comment:
Do norms make sense for this type of field? given we are searching within the object the length of the whole object being taken into account when searching might not be that interesting?

Reply:
Right, I think they would be misleading unless we had some way to filter them based on a particular key. I also suspect it's not common to use norms with `keyword` fields. I'll disallow norms for now, and potentially revisit if we have other ideas.